p2p/validation/MESSAGE_VALIDATION_CONSENSUS.MD

```diff
@@ -0,0 +1,455 @@
+3. Perhaps it will be better to have some past rounds ignored and not rejected, depending on how far in the past they are?
+
+
+#### Too many messages check
+
+This check ensures that once a certain message type was marked for a certain instance, round and height, then other messages for the same parameters can be safely ignored (signer mark) or rejected (peer mark).
+
+In the peer mark case, the mark should be updated on the fly.
+
+##### Code
+
+```go
+func (signerMark *mark) tooManyMsgsPerRound(round qbft.Round, msgType qbft.MessageType, share *ssvtypes.SSVShare) pubsub.ValidationResult {
+	if signerMark.MsgTypesInRound == nil {
+        signerMark.MsgTypesInRound = hashmap.New[qbft.Round, *hashmap.Map[qbft.MessageType, struct{}]]()
+	}
+	msgTypeToOccurrence, found := signerMark.MsgTypesInRound.Get(round)
+	if !found {
+		// must not initialize inner map here because memory allocation should happen only after signature validation
+		return pubsub.ValidationAccept
+	}
+	_, exists := msgTypeToOccurrences.Get(msgType)
+
+    if exists {
+        return pubsub.ValidationIgnore
+    }
+    return pubsub.ValidationAccept
+}
+
+func (peerMark *mark) tooManyMsgsPerRound(round qbft.Round, msgType qbft.MessageType, share *ssvtypes.SSVShare) pubsub.ValidationResult {
+	if peerMark.MsgTypesInRound == nil {
+        peerMark.MsgTypesInRound = hashmap.New[qbft.Round, *hashmap.Map[qbft.MessageType, struct{}]]()
+	}
+	msgTypeToOccurrence, found := peerMark.MsgTypesInRound.Get(round)
+	if !found {
+        // count on timely checks and decided updates to keep memory growth in check
+        msgTypeToOccurence = hasmap.New[qbft.MessageType, struct{}]
+        peerMark.MsgTypesInRound.Insert(round, msgTypeToOccurrence)
+	}
+	_, exists := msgTypeToOccurrences.Get(msgType)
+
+    if exists {
+        return pubsub.ValidationReject
+    }
+    msgTypeToOccurrences.Insert(msgType, struct{}{})
+    return pubsub.ValidationAccept
+}
+```
+
+
+### Signature Validation
+
+The final step before marking a message is to verify the signature. Since BLS signature validation is the expected computing bottleneck, a dedicated go routine performs [aggregate verification](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-bls-signature-04#section-3.2.3) on  *verification tuples* of < $signature, root, publicKey$ >, where:
+1. $signature=message.signature$
+2. $root=ComputeSigningRoot(message)$ ([See definition](https://github.com/bloxapp/ssv-spec/blob/2ae346065272800b061ffa4fde3aee062b5f4b45/types/crypto.go#L138-L146)).
+3. $public key=Aggregate(map(signer : message.signers).to (signer.getPublicKey()))$ $Aggregate$ is simply the addition of EC points as done in *[FastAggregateVerify](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-bls-signature-04#section-3.3.4)*.
+
+#### Algorithm
+
+1. Start a ticker that will create a tick event every `signatureVerificationInterval`
```

Comment:
not sure how you see aggregation working out... won't it delay execution? you'll be waiting on aggregation and not propagating the message..

Reply:
We wait at most 50 ms atm. Assuming a [small-world network](https://en.wikipedia.org/wiki/Small-world_network) bounded by 6 degrees of separation, then the max delay introduced by it will be 300 ms. In reality I believe the network diameter to be probably less than 6.

Prysm btw is doing exactly the same.

Reply:
This needs to be tested ... we are aggregating 2-3 messages in that time frame, I wonder if it's more efficient than individually verifying the message